tests/validate_tests.rs

```diff
@@ -0,0 +1,512 @@
+        "balance-1".to_string(),
+        None,
+        init_code,
+        init_func,
+        context.storage_factory.address,
+    )
+    .await
+    .expect_err("account has forbidden read");
+}
+
+#[tokio::test]
+async fn fail_if_referencing_self_token_balance_after_wallet_creation() {
+    test_existing_user_op("balance-self".to_string(), "".to_string())
+        .await
+        .expect("succeed");
+}
+
+#[tokio::test]
+#[should_panic]
+async fn fail_with_unstaked_paymaster_returning_context() {
+    let context = get_global_context().await;
+    let pm = deploy_test_storage_account(context.client.clone())
+        .await
+        .expect("deploy succeed");
+    let acct = deploy_test_recursion_account(context.client.clone(), context.entry_point.address)
+        .await
+        .expect("deploy succeed");
+
+    let mut paymaster_and_data = vec![];
+    paymaster_and_data.extend_from_slice(pm.address.as_bytes());
+    paymaster_and_data.extend_from_slice("postOp-context".as_bytes());
+
+    let user_op = UserOperation {
+        sender: acct.address,
+        nonce: U256::zero(),
+        init_code: Bytes::default(),
+        call_data: Bytes::default(),
+        call_gas_limit: U256::from(0),
+        verification_gas_limit: U256::from(50000),
+        pre_verification_gas: U256::from(0),
+        max_fee_per_gas: U256::from(0),
+        max_priority_fee_per_gas: U256::from(0),
+        paymaster_and_data: Bytes::from(paymaster_and_data),
+        signature: Bytes::default(),
+    };
+    validate(user_op).expect_err("unstaked paymaster must not return context");
+}
+
+#[tokio::test]
+#[should_panic]
+async fn fail_with_validation_recursively_calls_handle_ops() {
+    let context = get_global_context().await;
+    let acct = deploy_test_recursion_account(context.client.clone(), context.entry_point.address)
+        .await
+        .expect("deploy succeed");
+    let user_op = UserOperation {
+        sender: acct.address,
+        nonce: U256::zero(),
+        init_code: Bytes::default(),
+        call_data: Bytes::default(),
+        call_gas_limit: U256::from(0),
+        verification_gas_limit: U256::from(50000),
+        pre_verification_gas: U256::from(50000),
+        max_fee_per_gas: U256::from(0),
+        max_priority_fee_per_gas: U256::from(0),
+        paymaster_and_data: Bytes::default(),
+        signature: Bytes::from("handleOps".as_bytes().to_vec()),
+    };
+    validate(user_op).expect_err("illegal call into EntryPoint");
+}
+
+#[tokio::test]
+async fn succeed_with_inner_revert() {
+    let context = get_global_context().await;
+    let (init_code, init_func) = create_storage_factory_init_code(0, "".to_string())
+        .await
+        .unwrap();
+    test_user_op(
+        "inner-revert".to_string(),
+        None,
+        init_code,
+        init_func,
+        context.storage_factory.address,
+    )
+    .await
+    .expect("succeed");
+}
+
+#[tokio::test]
+#[should_panic]
+async fn fail_with_inner_oog_revert() {
```

Comment:
this test fails: 'test did not panic as expected'

Reply:
There are some underministic problem about the tests. I think it is caused by the singleton of the TestContext. I am looking at it.